rdkit/Chem/Draw/__init__.py

```diff
@@ -573,6 +571,238 @@ def _MolsToGridImage(mols, molsPerRow=3, subImgSize=(200, 200), legends=None,
   return res
 
 
+def _MolsNestedToLinear(mols_matrix, legends_matrix = None, highlightAtomLists_matrix = None,
+                        highlightBondLists_matrix = None):
+  """Converts a nested data structure (where each data substructure represents a row in mol grid image)
+  to a linear one, padding rows as needed so all rows are the length of the longest row
+        ARGUMENTS:
+
+        - mols_matrix: A two-deep nested data structure of RDKit molecules to draw, for example list of lists of RDKit molecules
+
+        - legends_matrix: A two-deep nested data structure of strings to label molecules with, for example list of lists of strings (default None)
+
+        - highlightAtomLists_matrix: A three-deep nested data structure of integers of atoms to highlight, for example list of lists of lists of integers (default None)
+
+        - highlightBondLists_matrix: A three-deep nested data structure of integers of bonds to highlight, for example list of lists of lists of integers (default None)
+
+      NOTE:
+
+            This private function does the heavy processing for MolsMatrixToGridImage
+
+      RETURNS:
+
+        Tuple of flattened (by one layer) data structures that MolsToGridImage accepts as input:
+        
+        - mols: A non-nested data structure of RDKit molecules to draw, for example list of RDKit molecules
+
+        - molsPerRow: The number of molecules per row (integer), equal to the longest row in mols_matrix
+
+        - legends: A non-nested data structure of strings to label molecules with, for example list of strings
+
+        - highlightAtomLists: A two-deep nested data structure of integers of atoms to highlight, for example list of lists of integers
+
+        - highlightBondLists: A two-deep nested data structure of integers of bonds to highlight, for example list of lists of integers
+
+  """
+  # Check that each item in nested lists is a list
+  def check_elements_are_lists(nested_list, nested_list_name=""):
```

Comment:
What's the value of this check? Shouldn't I also be able to provide a list of tuples, or tuple of lists, or tuple of tuples,  etc.? This is Python, there is *no* reason to enforce rigid typing.

I think the requirement here is that you need sequences which support the `len()` function and iteration

Reply:
Second comment here: all of these nested functions make it much harder to actually find the code for this function. Please pull them all out and put them at the global scope with names starting with `_` that are compatible with RDKit naming conventions.

Reply:
First comment: Great point, my intent was to verify that the elements are iterables, not specifically lists. I just changed to check that each element has the `__iter__` and `__len__` attributes.

Second comment: Will do; I debated doing that.

Reply:
> First comment: Great point, my intent was to verify that the elements are iterables, not specifically lists. I just changed to check that each element has the `__iter__` and `__len__` attributes.

But you don't need to do that explicitly since Python will do it for you and raise a TypeError with an informative message if the caller gets it wrong:
```
In [32]: def loopy(thing):
    ...:     res = [0]*len(thing)
    ...:     for i,item in enumerate(thing):
    ...:         res[i] = item+1
    ...:     return res
    ...: 

In [33]: loopy((1,2,3))
Out[33]: [2, 3, 4]

In [34]: loopy(123)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [34], line 1
----> 1 loopy(123)

Cell In [32], line 2, in loopy(thing)
      1 def loopy(thing):
----> 2     res = [0]*len(thing)
      3     for i,item in enumerate(thing):
      4         res[i] = item+1

TypeError: object of type 'int' has no len()
```

There's no sense in doing this check yourself, it's much better to let the language do the work for you and ensure that the same exceptions are thrown for the same types of errors.